### ISSUE TITLE
Windows: conda cannot activate environments with `!` in the name

### ISSUE THREAD

**Issue by marcoesters** [conda/conda#12558]
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

On Windows, [constructor](https://github.com/conda/constructor) disallows installing `conda` into paths that contain the special characters `^%!=,`: https://github.com/conda/constructor/blob/main/constructor/nsis/main.nsi.tmpl#L854

So, I got curious and tried to create conda environment with these:

```
conda create -n foo^%,!=
```

It created the environment `foo,!=`

```
# conda environments:                                                                                                                              
#   

base                  *  C:\miniconda3                                                                                                             
foo,!=                   C:\miniconda3\envs\foo,!= 
```

I verified that the directory exists. When I tried to activate this environment using `conda activate foo,!=`, I get the following error:

```                                                                                                                                                                                                     
EnvironmentNameNotFound: Could not find conda environment: foo,=                                                                                   
You can list all discoverable environments with `conda info --envs`.    
```

It seems like `!` is a problematic character, so `conda` should maybe not allow environments with that character on Windows. Maybe the list of forbidden characters should be synchronized with `constructor`, too?

### Conda Info

```shell
active environment : base                                                                                                                     
    active env location : C:\miniconda3                                                                                                            
            shell level : 1                                                                                                                        
       user config file : C:\Users\dev-admin\.condarc                                                                                              
 populated config files : C:\miniconda3\.condarc                                                                                                   
                          C:\Users\dev-admin\.condarc                                                                                              
          conda version : 23.3.1                                                                                                                   
    conda-build version : 3.24.0                                                                                                                   
         python version : 3.9.16.final.0                                                                                                           
       virtual packages : __archspec=1=x86_64                                                                                                      
                          __win=0=0                                                                                                                
       base environment : C:\miniconda3  (writable)                                                                                                
      conda av data dir : C:\miniconda3\etc\conda                                                                                                  
  conda av metadata url : None                                                                                                                     
           channel URLs : https://repo.anaconda.com/pkgs/main/win-64                                                                               
                          https://repo.anaconda.com/pkgs/main/noarch                                                                               
                          https://repo.anaconda.com/pkgs/r/win-64                                                                                  
                          https://repo.anaconda.com/pkgs/r/noarch                                                                                  
                          https://repo.anaconda.com/pkgs/msys2/win-64                                                                              
                          https://repo.anaconda.com/pkgs/msys2/noarch                                                                              
          package cache : C:\miniconda3\pkgs                                                                                                       
                          C:\Users\dev-admin\.conda\pkgs                                                                                           
                          C:\Users\dev-admin\AppData\Local\conda\conda\pkgs                                                                        
       envs directories : C:\miniconda3\envs                                                                                                       
                          C:\Users\dev-admin\.conda\envs                                                                                           
                          C:\Users\dev-admin\AppData\Local\conda\conda\envs                                                                        
               platform : win-64                                                                                                                   
             user-agent : conda/23.3.1 requests/2.28.1 CPython/3.9.16 Windows/10 Windows/10.0.17763                                                
          administrator : True                                                                                                                     
             netrc file : None                                                                                                                     
           offline mode : False
```


### Conda Config

```shell
==> C:\miniconda3\.condarc <==                                                                                                                     
add_pip_as_python_dependency: False                                                                                                                
show_channel_urls: True                                                                                                                            
                                                                                                                                                   
==> C:\Users\dev-admin\.condarc <==                                                                                                                
channels:                                                                                                                                          
  - defaults                                                                                                                                       
anaconda_upload: False
```


### Conda list

```shell
# packages in environment at C:\miniconda3:                                                                                                        
#                                                                                                                                                  
# Name                    Version                   Build  Channel                                                                                 
abs-sdk                   0.1.0+317.gc404253a            py_0    file:///C:/prefect/sdk                                                            
aiobotocore               2.4.0            py39haa95532_0    defaults                                                                              
aiohttp                   3.8.3            py39h2bbff1b_0    defaults                                                                              
aioitertools              0.7.1              pyhd3eb1b0_0    defaults                                                                              
aiosignal                 1.2.0              pyhd3eb1b0_0    defaults                                                                              
anaconda-client           1.11.1           py39haa95532_0    defaults                                                                              
appdirs                   1.4.4              pyhd3eb1b0_0    defaults                                                                              
asn1crypto                1.5.1            py39haa95532_0    defaults                                                                              
async-timeout             4.0.2            py39haa95532_0    defaults                                                                              
atomicwrites              1.4.0                      py_0    defaults                                                                              
attrs                     22.1.0           py39haa95532_0    defaults                                                                              
beautifulsoup4            4.11.1           py39haa95532_0    defaults                                                                              
boltons                   23.0.0           py39haa95532_0    defaults                                                                              
boto3                     1.24.28          py39haa95532_0    defaults                                                                              
botocore                  1.27.59          py39haa95532_0    defaults                                                                              
brotlipy                  0.7.0           py39h2bbff1b_1003    defaults                                                                            
bzip2                     1.0.8                he774522_0    defaults                                                                              
ca-certificates           2023.01.10           haa95532_0    defaults                                                                              
certifi                   2022.12.7        py39haa95532_0    defaults                                                                              
cffi                      1.15.1           py39h2bbff1b_3    defaults                                                                              
chardet                   4.0.0           py39haa95532_1003    defaults                                                                            
charset-normalizer        2.0.4              pyhd3eb1b0_0    defaults                                                                              
click                     8.0.4            py39haa95532_0    defaults                                                                              
cloudpickle               2.0.0              pyhd3eb1b0_0    defaults                                                                              
clyent                    1.2.2            py39haa95532_1    defaults                                                                              
colorama                  0.4.6            py39haa95532_0    defaults                                                                              
conda                     23.3.1           py39haa95532_0    defaults                                                                              
conda-build               3.24.0           py39haa95532_0    defaults                                                                              
conda-package-handling    1.9.0            py39h8cc25b3_0    defaults                                                                              
conda-package-streaming   0.7.0            py39haa95532_0    defaults                                                                              
console_shortcut          0.1.1                         4    defaults                                                                              
croniter                  0.3.35                     py_0    defaults                                                                              
cryptography              39.0.1           py39h21b164f_0    defaults                                                                              
dask-core                 2022.7.0         py39haa95532_0    defaults                                                                              
datadog                   0.42.0             pyhd3eb1b0_0    defaults                                                                              
defusedxml                0.7.1              pyhd3eb1b0_0    defaults                                                                              
deprecated                1.2.13           py39haa95532_0    defaults                                                                              
distributed               2022.7.0         py39haa95532_0    defaults                                                                              
docker-py                 4.4.1            py39haa95532_5    defaults                                                                              
docker-pycreds            0.4.0              pyhd3eb1b0_0    defaults                                                                              
filelock                  3.9.0            py39haa95532_0    defaults                                                                              
flit-core                 3.8.0            py39haa95532_0    defaults                                                                              
frozenlist                1.3.3            py39h2bbff1b_0    defaults                                                                              
fsspec                    2023.3.0         py39haa95532_0    defaults                                                                              
gitdb                     4.0.7              pyhd3eb1b0_0    defaults                                                                              
gitpython                 3.1.30           py39haa95532_0    defaults                                                                              
glob2                     0.7                pyhd3eb1b0_0    defaults                                                                              
heapdict                  1.0.1              pyhd3eb1b0_0    defaults                                                                              
idna                      3.4              py39haa95532_0    defaults                                                                              
importlib_resources       5.2.0              pyhd3eb1b0_1    defaults                                                                              
iniconfig                 1.1.1              pyhd3eb1b0_0    defaults                                                                              
jinja2                    3.1.2            py39haa95532_0    defaults                                                                              
jmespath                  0.10.0             pyhd3eb1b0_0    defaults                                                                              
jsonpatch                 1.32               pyhd3eb1b0_0    defaults                                                                              
jsonpointer               2.1                pyhd3eb1b0_0    defaults                                                                              
jsonschema                4.17.3           py39haa95532_0    defaults                                                                              
jupyter_core              5.2.0            py39haa95532_0    defaults                                                                              
libarchive                3.6.2                h2033e3e_1    defaults                                                                              
libiconv                  1.16                 h2bbff1b_2    defaults                                                                              
liblief                   0.12.3               hd77b12b_0    defaults                                                                              
libxml2                   2.10.3               h0ad7f3c_0    defaults                                                                              
locket                    1.0.0            py39haa95532_0    defaults                                                                              
lz4-c                     1.9.4                h2bbff1b_0    defaults                                                                              
m2-base                   1.0.0                         3    defaults                                                                              
m2-bash                   4.3.042                       5    defaults                                                                              
m2-bash-completion        2.3                           2    defaults                                                                              
m2-catgets                1.1                           3    defaults                                                                              
m2-coreutils              8.25                        102    defaults                                                                              
m2-dash                   0.5.8                         2    defaults                                                                              
m2-diffutils              3.3                           4    defaults                                                                              
m2-file                   5.25                          2    defaults                                                                              
m2-filesystem             2016.04                       4    defaults                                                                              
m2-findutils              4.6.0                         2    defaults                                                                              
m2-gawk                   4.1.3                         2    defaults                                                                              
m2-gcc-libs               5.3.0                         4    defaults                                                                              
m2-gettext                0.19.7                        4    defaults                                                                              
m2-gmp                    6.1.0                         3    defaults                                                                              
m2-grep                   2.22                          4    defaults                                                                              
m2-gzip                   1.7                           2    defaults                                                                              
m2-inetutils              1.9.2                         2    defaults                                                                              
m2-info                   6.0                           2    defaults                                                                              
m2-less                   481                           2    defaults                                                                              
m2-libasprintf            0.19.7                        4    defaults                                                                              
m2-libbz2                 1.0.6                         3    defaults                                                                              
m2-libcatgets             1.1                           3    defaults                                                                              
m2-libcrypt               1.3                           2    defaults                                                                              
m2-libgettextpo           0.19.7                        4    defaults                                                                              
m2-libiconv               1.14                          3    defaults                                                                              
m2-libintl                0.19.7                        4    defaults                                                                              
m2-liblzma                5.2.2                         2    defaults                                                                              
m2-libpcre                8.38                          2    defaults                                                                              
m2-libreadline            6.3.008                       8    defaults                                                                              
m2-libutil-linux          2.26.2                        2    defaults                                                                              
m2-libxml2                2.9.2                         3    defaults                                                                              
m2-make                   4.1                           5    defaults                                                                              
m2-mintty                 1!2.2.3                       2    defaults                                                                              
m2-mpfr                   3.1.4                         2    defaults                                                                              
m2-msys2-launcher-git     0.3.28.860c495                2    defaults                                                                              
m2-msys2-runtime          2.5.0.17080.65c939c               3    defaults                                                                          
m2-ncurses                6.0.20160220                  2    defaults                                                                              
m2-patch                  2.7.5                         2    defaults                                                                              
m2-sed                    4.2.2                         3    defaults                                                                              
m2-tar                    1.28                          4    defaults                                                                              
m2-tftp-hpa               5.2                           2    defaults                                                                              
m2-time                   1.7                           2    defaults                                                                              
m2-ttyrec                 1.0.8                         2    defaults                                                                              
m2-tzcode                 2015.e                        2    defaults                                                                              
m2-unzip                  6.0                           3    defaults                                                                              
m2-util-linux             2.26.2                        2    defaults                                                                              
m2-which                  2.21                          3    defaults                                                                              
m2-zip                    3.0                           2    defaults                                                                              
m2-zlib                   1.2.8                         4    defaults                                                                              
markupsafe                2.1.1            py39h2bbff1b_0    defaults                                                                              
marshmallow               3.12.2             pyhd3eb1b0_0    defaults                                                                              
marshmallow-oneofschema   3.0.1            py39haa95532_0    defaults                                                                              
menuinst                  1.4.19           py39h59b6b97_0    defaults                                                                              
msgpack-python            1.0.3            py39h59b6b97_0    defaults                                                                              
msys2-conda-epoch         20160418                      1    defaults                                                                              
multidict                 6.0.2            py39h2bbff1b_0    defaults                                                                              
mypy_extensions           0.4.3            py39haa95532_1    defaults                                                                              
natsort                   7.1.1              pyhd3eb1b0_0    defaults                                                                              
nbformat                  5.7.0            py39haa95532_0    defaults                                                                              
openssl                   1.1.1t               h2bbff1b_0    defaults                                                                              
packaging                 23.0             py39haa95532_0    defaults                                                                              
partd                     1.2.0              pyhd3eb1b0_1    defaults                                                                              
pendulum                  2.1.2              pyhd3eb1b0_1    defaults                                                                              
pip                       23.0.1           py39haa95532_0    defaults                                                                              
pkginfo                   1.9.6            py39haa95532_0    defaults                                                                              
platformdirs              2.5.2            py39haa95532_0    defaults                                                                              
pluggy                    1.0.0            py39haa95532_1    defaults                                                                              
posix                     1.0.0                         2    defaults                                                                              
powershell_shortcut       0.0.1                         3    defaults                                                                              
prefect                   1.4.0              pyhd3eb1b0_0    distro-tooling                                                                        
prefect-cookbook          0+unknown                pypi_0    pypi                                                                                  
psutil                    5.9.0            py39h2bbff1b_0    defaults                                                                              
py                        1.11.0             pyhd3eb1b0_0    defaults                                                                              
py-lief                   0.12.3           py39hd77b12b_0    defaults                                                                              
pycosat                   0.6.4            py39h2bbff1b_0    defaults                                                                              
pycparser                 2.21               pyhd3eb1b0_0    defaults                                                                              
pydantic                  1.10.2           py39h2bbff1b_0    defaults                                                                              
pygithub                  1.55               pyhd3eb1b0_1    defaults                                                                              
pyjwt                     2.4.0            py39haa95532_0    defaults                                                                              
pynacl                    1.5.0            py39h8cc25b3_0    defaults                                                                              
pyopenssl                 23.0.0           py39haa95532_0    defaults                                                                              
pyparsing                 3.0.9            py39haa95532_0    defaults                                                                              
pyrsistent                0.18.0           py39h196d8e1_0    defaults                                                                              
pysocks                   1.7.1            py39haa95532_0    defaults                                                                              
pytest                    7.1.2            py39haa95532_0    defaults                                                                              
python                    3.9.16               h6244533_2    defaults                                                                              
python-box                5.4.1              pyhd3eb1b0_0    distro-tooling                                                                        
python-dateutil           2.8.2              pyhd3eb1b0_0    defaults                                                                              
python-fastjsonschema     2.16.2           py39haa95532_0    defaults                                                                              
python-json-logger        2.0.1                      py_0    defaults                                                                              
python-libarchive-c       2.9                pyhd3eb1b0_1    defaults                                                                              
python-slugify            5.0.2              pyhd3eb1b0_0    defaults                                                                              
pytz                      2022.7           py39haa95532_0    defaults                                                                              
pytzdata                  2020.1             pyhd3eb1b0_0    defaults                                                                              
pywin32                   305              py39h2bbff1b_0    defaults                                                                              
pyyaml                    6.0              py39h2bbff1b_1    defaults                                                                              
requests                  2.28.1           py39haa95532_1    defaults                                                                              
ruamel.yaml               0.17.21          py39h2bbff1b_0    defaults                                                                              
ruamel.yaml.clib          0.2.6            py39h2bbff1b_1    defaults                                                                              
ruamel_yaml               0.15.100         py39h2bbff1b_0    defaults                                                                              
s3transfer                0.6.0            py39haa95532_0    defaults                                                                              
setuptools                65.6.3           py39haa95532_0    defaults                                                                              
six                       1.16.0             pyhd3eb1b0_1    defaults                                                                              
slack-sdk                 3.19.5             pyhaa95532_0    distro-tooling                                                                        
smmap                     4.0.0              pyhd3eb1b0_0    defaults                                                                              
sortedcontainers          2.4.0              pyhd3eb1b0_0    defaults                                                                              
soupsieve                 2.3.2.post1      py39haa95532_0    defaults                                                                              
sqlite                    3.41.1               h2bbff1b_0    defaults                                                                              
tabulate                  0.8.10           py39haa95532_0    defaults                                                                              
tblib                     1.7.0              pyhd3eb1b0_0    defaults                                                                              
text-unidecode            1.3                pyhd3eb1b0_0    defaults                                                                              
toml                      0.10.2             pyhd3eb1b0_0    defaults                                                                              
tomli                     2.0.1            py39haa95532_0    defaults                                                                              
toolz                     0.12.0           py39haa95532_0    defaults                                                                              
tornado                   6.1              py39h2bbff1b_0    defaults                                                                              
tqdm                      4.65.0           py39hd4e2768_0    defaults                                                                              
traitlets                 5.7.1            py39haa95532_0    defaults                                                                              
typing-extensions         4.4.0            py39haa95532_0    defaults                                                                              
typing_extensions         4.4.0            py39haa95532_0    defaults                                                                              
tzdata                    2022g                h04d1e81_0    defaults                                                                              
unidecode                 1.2.0              pyhd3eb1b0_0    defaults                                                                              
urllib3                   1.26.14          py39haa95532_0    defaults                                                                              
vc                        14.2                 h21ff451_1    defaults                                                                              
vs2015_runtime            14.27.29016          h5e58377_2    defaults                                                                              
websocket-client          0.58.0           py39haa95532_4    defaults                                                                              
wheel                     0.38.4           py39haa95532_0    defaults                                                                              
win_inet_pton             1.1.0            py39haa95532_0    defaults                                                                              
wincertstore              0.2              py39haa95532_2    defaults                                                                              
wrapt                     1.14.1           py39h2bbff1b_0    defaults                                                                              
xz                        5.2.10               h8cc25b3_1    defaults                                                                              
yaml                      0.2.5                he774522_0    defaults                                                                              
yarl                      1.8.1            py39h2bbff1b_0    defaults                                                                              
zict                      2.1.0            py39haa95532_0    defaults                                                                              
zipp                      3.11.0           py39haa95532_0    defaults                                                                              
zlib                      1.2.13               h8cc25b3_0    defaults                                                                              
zstandard                 0.19.0           py39h2bbff1b_0    defaults                                                                              
zstd                      1.5.2                h19a0ad4_0    defaults
```


### Additional Context

_No response_

**Comment by marcoesters** [comment]
@jaimergp and I did some poking around and it looks like the culprit is inside the `_conda_activate.bat` file.

The file sets `EnableDelayedExpansion`, which causes the `!` to be dropped from the environment name (I echoed out `%*` before and after to confirm this).

The easiest solution would be to prohibit this character when creating an environment. Getting allowed characters synced with the constructor repo is probably generally not a bad idea.

**Comment by jaimergp** [comment]
Still an issue:

https://github.com/conda/conda/blob/4a4aa7732f7bcbbcbb85ac197d32e83254cc0f32/conda/base/constants.py#L181-L182

**Comment by mbargull** [comment]
I think it makes sense to avoid/discourage the usage of additional character like `!` (which can also be problematic for people on (interactive) Unix shells to use, FYI).

The changes from gh-13975 might cause more harm than good, though.
In recent times, we tried to be more diligent wrt. to deprecation cycles, but the change from gh-13975 is a direct user-facing change with no further notice that happened beforehand.
It did not only disallow `!` (which the user would've found to be non-usable due to this issue here), but also other characters like `=`, `(`, `)`, which are not unlikely to be chosen by users (e.g., `conda create --name "python=3.12" python=3.12` seems reasonable; `conda create --name "myenv(new)"` might even be a thing too).

This means, that change is likely to break some users' workflows that create environment with such patterns/characters.

Additionally, gh-13975 did not carry any additional test to ensure existing environments are not affected.
I have yet to try it out, but I believe that check is also used for existing environment names, e.g., a `python=3.12`-named environment created by `conda<24.7` might not be serviceable in `conda=24.7.0`  anymore due to that change.

----
I'd advice to
1. Revert gh-13975 for now and reopen this issue,
2. Fix the activation script to handle `!` (I can do that; should be not a big problem),
3. Write some more tests to cover handling of existing environments,

and after that then reevaluate how/where to implement disallowing/discouraging uses of certain characters in environment names going forward.

(cc @beeankha, @jezdez)

**Comment by jaimergp** [comment]
Thanks @mbargull. Added https://github.com/conda/conda/pull/14065.

**Comment by mbargull** [comment]
Thanks @jaimergp  (sorry, for the timing :pensive:).

I tried it out in a VM
<sub>(Well, after Linux' OOM killer decided to kill *everything* I had opened *but* that VM :vomiting_face: .)</sub>

Looks like some thing do work but some don't before/after the change;
E.g., `conda install --name=python=3.12 zstd` does not work afterwards but `conda activate python=3.12` still does.
(The prompt-handling on Windows doesn't seem to like the `=` in any case, apparently.)
```cmd
C:\conda\base-from-24.5.0>condabin\conda activate

(base) C:\conda\base-from-24.5.0>conda info

     active environment : base
    active env location : C:\conda\base-from-24.5.0
            shell level : 1
       user config file : C:\Users\conda\.condarc
 populated config files :
          conda version : 24.5.0
    conda-build version : not installed
         python version : 3.12.4.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=x86_64_v4
                          __conda=24.5.0=0
                          __win=0=0
       base environment : C:\conda\base-from-24.5.0  (writable)
      conda av data dir : C:\conda\base-from-24.5.0\etc\conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/win-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/win-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/msys2/win-64
                          https://repo.anaconda.com/pkgs/msys2/noarch
          package cache : C:\conda\base-from-24.5.0\pkgs
                          C:\Users\conda\.conda\pkgs
                          C:\Users\conda\AppData\Local\conda\conda\pkgs
       envs directories : C:\conda\base-from-24.5.0\envs
                          C:\Users\conda\.conda\envs
                          C:\Users\conda\AppData\Local\conda\conda\envs
               platform : win-64
             user-agent : conda/24.5.0 requests/2.32.3 CPython/3.12.4 Windows/11 Windows/10.0.22631 solver/libmamba conda-libmamba-solver/24.7.0 libmambapy/1.5.8
          administrator : False
             netrc file : None
           offline mode : False



(base) C:\conda\base-from-24.5.0>conda create --name=python=3.12 python=3.12
Channels:
 - defaults
Platform: win-64
Collecting package metadata (repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: C:\conda\base-from-24.5.0\envs\python=3.12

  added / updated specs:
    - python=3.12


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    bzip2-1.0.8                |       h2bbff1b_6          90 KB
    ca-certificates-2024.7.2   |       haa95532_0         128 KB
    expat-2.6.2                |       hd77b12b_0         260 KB
    libffi-3.4.4               |       hd77b12b_1         122 KB
    openssl-3.0.14             |       h827c3e9_0         7.8 MB
    pip-24.0                   |  py312haa95532_0         3.5 MB
    python-3.12.4              |       h14ffc60_1        16.7 MB
    setuptools-69.5.1          |  py312haa95532_0         1.3 MB
    sqlite-3.45.3              |       h2bbff1b_0         973 KB
    tk-8.6.14                  |       h0416ee5_0         3.5 MB
    tzdata-2024a               |       h04d1e81_0         116 KB
    vc-14.2                    |       h2eaa2aa_4          10 KB
    vs2015_runtime-14.29.30133 |       h43f2093_4         1.1 MB
    wheel-0.43.0               |  py312haa95532_0         167 KB
    xz-5.4.6                   |       h8cc25b3_1         609 KB
    zlib-1.2.13                |       h8cc25b3_1         131 KB
    ------------------------------------------------------------
                                           Total:        36.4 MB

The following NEW packages will be INSTALLED:

  bzip2              pkgs/main/win-64::bzip2-1.0.8-h2bbff1b_6
  ca-certificates    pkgs/main/win-64::ca-certificates-2024.7.2-haa95532_0
  expat              pkgs/main/win-64::expat-2.6.2-hd77b12b_0
  libffi             pkgs/main/win-64::libffi-3.4.4-hd77b12b_1
  openssl            pkgs/main/win-64::openssl-3.0.14-h827c3e9_0
  pip                pkgs/main/win-64::pip-24.0-py312haa95532_0
  python             pkgs/main/win-64::python-3.12.4-h14ffc60_1
  setuptools         pkgs/main/win-64::setuptools-69.5.1-py312haa95532_0
  sqlite             pkgs/main/win-64::sqlite-3.45.3-h2bbff1b_0
  tk                 pkgs/main/win-64::tk-8.6.14-h0416ee5_0
  tzdata             pkgs/main/noarch::tzdata-2024a-h04d1e81_0
  vc                 pkgs/main/win-64::vc-14.2-h2eaa2aa_4
  vs2015_runtime     pkgs/main/win-64::vs2015_runtime-14.29.30133-h43f2093_4
  wheel              pkgs/main/win-64::wheel-0.43.0-py312haa95532_0
  xz                 pkgs/main/win-64::xz-5.4.6-h8cc25b3_1
  zlib               pkgs/main/win-64::zlib-1.2.13-h8cc25b3_1


Proceed ([y]/n)?


Downloading and Extracting Packages:

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate python=3.12
#
# To deactivate an active environment, use
#
#     $ conda deactivate


(base) C:\conda\base-from-24.5.0>conda activate python=3.12

(python=3.12) C:\conda\base-from-24.5.0>python --version
Python 3.12.4

(python=3.12) C:\conda\base-from-24.5.0>conda activate base

(base) 3.12) ==3.12) C:\conda\base-from-24.5.0>conda install conda=24.7.0 -c conda-canary/label/rc-conda-24.7.x
Channels:
 - conda-canary/label/rc-conda-24.7.x
 - defaults
Platform: win-64
Collecting package metadata (repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: C:\conda\base-from-24.5.0

  added / updated specs:
    - conda=24.7.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    conda-24.7.0+0_gd486f898e  |          py312_0         1.2 MB  conda-canary/label/rc-conda-24.7.x
    ------------------------------------------------------------
                                           Total:         1.2 MB

The following packages will be UPDATED:

  conda              pkgs/main::conda-24.5.0-py312haa95532~ --> conda-canary/label/rc-conda-24.7.x::conda-24.7.0+0_gd486f898e-py312_0


Proceed ([y]/n)?


Downloading and Extracting Packages:

Preparing transaction: done
Verifying transaction: done
Executing transaction: done

(base) 3.12) ==3.12) C:\conda\base-from-24.5.0>conda install --name=python=3.12 zstd

CondaValueError: Invalid environment name: 'python=3.12'
Characters not allowed: {'\\', '/', ':', '=', '%', '^', '(', ' ', ')', '!', '#'}
If you are specifying a path to an environment, the `-p`
flag should be used instead.



(base) 3.12) ==3.12) C:\conda\base-from-24.5.0>conda activate python=3.12

(python=3.12) 3.12) ==3.12) C:\conda\base-from-24.5.0>conda list
# packages in environment at C:\conda\base-from-24.5.0\envs\python=3.12:
#
# Name                    Version                   Build  Channel
bzip2                     1.0.8                h2bbff1b_6
ca-certificates           2024.7.2             haa95532_0
expat                     2.6.2                hd77b12b_0
libffi                    3.4.4                hd77b12b_1
openssl                   3.0.14               h827c3e9_0
pip                       24.0            py312haa95532_0
python                    3.12.4               h14ffc60_1
setuptools                69.5.1          py312haa95532_0
sqlite                    3.45.3               h2bbff1b_0
tk                        8.6.14               h0416ee5_0
tzdata                    2024a                h04d1e81_0
vc                        14.2                 h2eaa2aa_4
vs2015_runtime            14.29.30133          h43f2093_4
wheel                     0.43.0          py312haa95532_0
xz                        5.4.6                h8cc25b3_1
zlib                      1.2.13               h8cc25b3_1

(python=3.12) 3.12) ==3.12) C:\conda\base-from-24.5.0>conda install zstd
Channels:
 - defaults
Platform: win-64
Collecting package metadata (repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: C:\conda\base-from-24.5.0\envs\python=3.12

  added / updated specs:
    - zstd


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    lz4-c-1.9.4                |       h2bbff1b_1         152 KB
    zstd-1.5.5                 |       hd43e919_2         720 KB
    ------------------------------------------------------------
                                           Total:         872 KB

The following NEW packages will be INSTALLED:

  lz4-c              pkgs/main/win-64::lz4-c-1.9.4-h2bbff1b_1
  zstd               pkgs/main/win-64::zstd-1.5.5-hd43e919_2


Proceed ([y]/n)?


Downloading and Extracting Packages:

Preparing transaction: done
Verifying transaction: done
Executing transaction: done

(python=3.12) 3.12) ==3.12) 3.12) ==3.12) C:\conda\base-from-24.5.0>
```

**Comment by jaimergp** [comment]
I see. Most of these issues will be CMD related I assume, so maybe they do work with Powershell and whatnot. An extremely delicate part of `conda` that does deserve better attention than the one I gave in the linked PR. So I agree that we should revert and cut 24.7.1 for now.

Maybe we need to "deprecate allowed characters" with a warning first, and then disallow completely. Maybe this is also part of a CEP where we explicitly define which environment names and paths are allowed. To be seen!

**Comment by jezdez** [comment]
Hey all, this is out of the release policy because the release had already been tagged and the feedback basically given too late.

As such the change shouldn’t have been reverted right away given the presented above. The reproducer above is pretty clear, but an edge case nonetheless.

I’d suggest to work on an improved patch that only presents users with a note that future versions of conda will change env name handling.

**Comment by jaimergp** [comment]
> As such the change shouldn’t have been reverted right away given the presented above.

FWIW, I think the revert PR has not been merged yet.

I've taken a quick look at Github to see how edge casey this might be and I found:

- Some [documentation](https://github.com/search?q=%2Fconda+create+-n+%5CS%2B%3D%5CS%2B%2F&ref=opensearch&type=code) instructing to `conda create -n python=x.y`
- [_Two_ `environment.yml`](https://github.com/search?q=%2F%5Ename%3A.*%3D.*%2F+path%3Aenvironment.yml+&ref=opensearch&type=code) (though from the same user) with a similar name

So not super common but also not unheard of.

**Comment by h-vetinari** [comment]
From the peanut gallery: this is IMO a rare edge case where people will just move on.

That said, I agree with the principle of deprecating potentially working patterns before breaking them, but vice versa, things that are already broken should just be forbidden directly (e.g. `!`).

> Looks like some things do work but some don't before/after the change;

More to the point, anything that already produces results like
```
(python=3.12) 3.12) ==3.12) 3.12) ==3.12) C:\conda\base-from-24.5.0>
```
from basic use should just be considered broken and is IMO not worth the deprecation dance for something so dysfunctional (which will thus have close to zero users).

If there are characters that allow barebones usage (e.g. creation, activation, deactivation) without obvious errors, then those should be deprecated, but I think it makes sense to draw a line of what constitutes viable existing usability (e.g. `=` fails).

**Comment by mbargull** [comment]
@h-vetinari, reopening this issue was done in context of potential breakage of users' existing workflows/environments -- a topic we have to handle with diligence.
The examples you mentioned are due to multiple bugs in one part of `conda`'s current code.
Although an important part, CMD shell handling is not a necessary functionality for all users -- there are multiple ways in which those bugs might not affect users (different shells, configuration or just differnt ways of invoking `conda`).
To be clear: The "deprecation dance" would absolutely necessary here since we are directly affecting end users and we are not talking about dysfunctional use cases.

Let us stay on topic on how to fix bugs and handle it in the broader context with the facts we have and let's avoid looking at narrower contexts stemming just from assumptions about users' workflows.

**Comment by h-vetinari** [comment]
I've handled a large amount of deprecations in scipy of all shapes and sizes, so I have some experience with this process.

Of course deprecation is preferable, though resources are not infinite, and if the cases being deprecated are already failing _and_ likely very rare, it's IMO a bad tradeoff to invest too much time on them.

As an open source maintainer, cf/core member and windows user, I think the effort here should be minimised (did it work before? yes->deprecate, no->forbid), but ultimately you all get to decide how you spend your time.

**Comment by kenodegard** [comment]
I appreciate everyone's insight here.

Since there are concerns about environment breakage with #13975 I have decided to revert the change in `conda 24.7.1` (see https://github.com/conda/conda/pull/14065).

The plan is to take stock of all prefix validation and verification efforts (e.g., this, https://github.com/conda/conda/pull/13955, https://github.com/conda/conda/issues/14042, etc.) and propose a solution for `conda 24.9.x`.

**Comment by travishathaway** [comment]
@kenodegard and @marcoesters,

Is this still a priority to get fixed? I can't help but noticed it's been repeatedly moved from release to release.

**Comment by jezdez** [comment]
## Proposed Plan for Fixing Issue #12558

After reviewing the history of PR #13975 (merged then reverted in #14065), here's a proposed approach that addresses the concerns raised.

### Background

The original fix was reverted because:
1. **Broke existing environments** - Envs with names like `python=3.12` became unserviceable via `--name` flag
2. **No deprecation cycle** - Direct breaking change with no warning period
3. **Over-broad scope** - Blocked `=`, `(`, `)` which are more commonly used than `!`
4. **Inconsistent behavior** - `conda activate python=3.12` worked but `conda install --name=python=3.12 zstd` didn't
5. **No tests for existing envs** - No validation that pre-existing environments would continue to work

### Good News

The batch script was **rewritten in PR #14607** - it now uses a static INI-style environment file approach instead of dynamic script generation. The `_conda_activate.bat` no longer uses `EnableDelayedExpansion`, which may have resolved the `!` activation issue.

### Proposed Character Categories

| Category | Characters | Current Behavior | Recommendation |
|----------|------------|------------------|----------------|
| **Completely Broken** | `!` | Can't activate in CMD (historically) | Disallow immediately |
| **Partially Broken** | `=`, `^`, `%` | Prompt corruption, some operations fail | Deprecation warning → disallow later |
| **Potentially Problematic** | `(`, `)`, `\` | May work but problematic in scripts | Deprecation warning |
| **Already Disallowed** | `/`, ` `, `:`, `#` | Currently blocked | Keep blocked |

### Proposed Rollout

#### Phase 1: Characterization (PR incoming)
- Add comprehensive test cases to characterize actual behavior on Windows
- Verify which characters are truly broken vs cosmetically problematic
- Test if PR #14607's changes fixed the `!` activation issue

#### Phase 2: Immediate Block (Release N)
- Block only `!` on Windows if still broken (completely broken, no working use case)
- Ensure existing `!` envs can still be accessed by `-p`/`--prefix` path

#### Phase 3: Deprecation Warnings (Release N)
- Add `FutureWarning` for `=`, `^`, `%`, `(`, `)` on Windows when creating new envs
- Message: "Environment names containing {char} will be disallowed in a future release"

#### Phase 4: Full Block (Release N+2 or N+3)
- Block all problematic characters for new env creation
- Existing envs remain accessible via `-p`/`--prefix`

### Key Principle

**Creating new environments** with problematic characters should be blocked/warned, but **existing environments** must remain usable (at minimum via the `-p` flag).

---

I'll be opening a PR shortly with test cases to characterize the actual behavior and validate the approach.

cc @kenodegard @marcoesters @jaimergp @mbargull

**Comment by jezdez** [comment]
## Status Update on #12558

I've been working on characterizing and fixing the special character issues. Here's the current status:

### Test Results from CI

| Character | Activation Test | Notes |
|-----------|-----------------|-------|
| `!` (exclamation) | ✅ **WORKS** | Fixed by PR #14607's batch script rewrite (removed `EnableDelayedExpansion`) |
| `^` (caret) | ✅ **FIXED** | Was broken - caret consumed as escape char. Fixed in PR #15628 |
| `=` (equals) | 🔄 Testing | Activation works, prompt may have issues |
| `%` (percent) | 🔄 Testing | Environment variable expansion concerns |
| `(` `)` (parens) | 🔄 Testing | Special CMD meaning |

### What's Been Done

1. **Comprehensive test suite added** (PR #15628) - Integration tests that actually activate environments with special characters on Windows CMD.EXE

2. **Fix for `^` character** - The caret was being consumed as CMD.EXE's escape character during batch script processing. Fixed by escaping `^` to `^^` in the INI file values.

3. **Confirmed `!` is fixed** - PR #14607's rewrite of `_conda_activate.bat` to use static INI files instead of dynamic script generation with `EnableDelayedExpansion` resolved the original `!` issue.

### Next Steps

1. Wait for full CI results on remaining characters (`=`, `%`, `(`, `)`)
2. Determine which characters need similar escaping fixes vs deprecation warnings
3. Implement deprecation path for characters that can't be fully fixed

### PR

See PR #15628 for the tests and fix: https://github.com/conda/conda/pull/15628

**Comment by jezdez** [comment]
## Follow-up: Fixed the escaping implementation

The previous commit had the escaping logic in `CmdExeActivator.template_export_var()` and `template_path_var()`, but these methods weren't being called because `_yield_commands()` was using the template strings directly instead of calling the template methods.

**Fix in commit 7b470365a**: Modified `_yield_commands()` to use `self.template_path_var(key, value)` instead of `self.export_var_tmpl % (key, value)`, etc. This ensures `CmdExeActivator`'s escaping logic is actually applied.

CI should now pass for the `^` character test.

**Comment by jezdez** [comment]
## Bug fix: Integer handling

The previous commit broke tests because `CONDA_SHLVL` is an integer, and `_escape_cmd_special_chars()` was calling `.replace()` on it directly:

```
AttributeError: 'int' object has no attribute 'replace'
```

Fixed in commit d9e3f9c04 by converting value to string first.

**Comment by jezdez** [comment]
## Final Status Update on #12558

All CI tests pass. Here's the final status:

### Test Results

| Character | Activation | Prompt | Status |
|-----------|------------|--------|--------|
| `!` (exclamation) | ✅ Pass | ✅ Pass | **WORKS** - Fixed by PR #14607 |
| `^` (caret) | ✅ Pass | ✅ Pass | **FIXED** - Fixed in PR #15628 |
| `=` (equals) | ✅ Pass | ✅ Pass | **WORKS** |
| `%` (percent) | ✅ Pass | ✅ Pass | **WORKS** |
| `(` (open paren) | ✅ Pass | ✅ Pass | **WORKS** |
| `)` (close paren) | ✅ Pass | ✅ Pass | **WORKS** |

### Conclusion

**All previously problematic characters now work correctly on Windows CMD.EXE.**

The original issue (`!` can't activate) was fixed by PR #14607's rewrite of `_conda_activate.bat` to use static INI-style files instead of dynamic script generation with `EnableDelayedExpansion`.

During characterization testing, we discovered `^` was still broken (consumed as CMD.EXE escape character) and fixed it in PR #15628 by escaping `^` to `^^`.

### Recommendation

Since all characters now work correctly:
- **No deprecation warnings needed** - All characters function properly
- **No character blocking needed** - Users can continue using these characters
- **PR #15628 ready for merge** - Contains comprehensive tests + caret fix

### What PR #15628 Delivers

1. **Bug fix**: Caret (`^`) character now works in environment names
2. **Test coverage**: Comprehensive integration tests for all 6 special characters
3. **Documentation**: Tests serve as living documentation of expected behavior
4. **Regression protection**: Future changes won't break special character support

PR: https://github.com/conda/conda/pull/15628

**Comment by jezdez** [comment]
Moving this to the 26.3 (March) release. The amount of debugging and iteration on PR #15628 spiraled a bit more than expected, and given the past experience with having to roll back #13975, it's too risky to merge this last minute for 26.1.

The fix is solid and well-tested, but we want to give it proper time in the release cycle to catch any edge cases.

**Comment by jezdez** [comment]
Quick update on PR #15628:

Addressed Ryan's feedback in 373e92004 - added escaping for activate/deactivate script paths as well. If the environment prefix contains a caret (e.g. `test^env`), the script paths in `etc/conda/activate.d/` would also contain it and need the same escaping treatment.

The PR now escapes special characters in all template methods for CmdExeActivator:
- `template_export_var`
- `template_path_var`
- `template_set_var`
- `template_run_script` (newly added)

Still waiting on CI and reviews.

**Comment by waqasanjum6** [comment]
@jezdez Im seeing below issue while verifying the fix on conda 26.1.1.65+g78a3de60c

```
C:\Users\WaqasAnjum>conda create -p "C:\envs\test^env" python=3.12 -y
3 channel Terms of Service accepted
Channels:
 - conda-forge
 - defaults
Platform: win-64
Collecting package metadata (repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: C:\envs\test^^env

  added / updated specs:
    - numpy
    - pip
    - python=3.12
    - requests


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    requests-2.33.1            |     pyhcf101f3_0          62 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          62 KB

The following NEW packages will be INSTALLED:

  backports.zstd     conda-forge/win-64::backports.zstd-1.3.0-py312h06d0912_0
  brotli-python      conda-forge/win-64::brotli-python-1.2.0-py312hc6d9e41_1
  bzip2              conda-forge/win-64::bzip2-1.0.8-h0ad9c76_9
  ca-certificates    conda-forge/noarch::ca-certificates-2026.2.25-h4c7d964_0
  certifi            conda-forge/noarch::certifi-2026.2.25-pyhd8ed1ab_0
  charset-normalizer conda-forge/noarch::charset-normalizer-3.4.6-pyhd8ed1ab_0
  h2                 conda-forge/noarch::h2-4.3.0-pyhcf101f3_0
  hpack              conda-forge/noarch::hpack-4.1.0-pyhd8ed1ab_0
  hyperframe         conda-forge/noarch::hyperframe-6.1.0-pyhd8ed1ab_0
  idna               conda-forge/noarch::idna-3.11-pyhd8ed1ab_0
  libblas            conda-forge/win-64::libblas-3.11.0-6_hf2e6a31_mkl
  libcblas           conda-forge/win-64::libcblas-3.11.0-6_h2a3cdd5_mkl
  libexpat           conda-forge/win-64::libexpat-2.7.5-hac47afa_0
  libffi             conda-forge/win-64::libffi-3.5.2-h3d046cb_0
  libhwloc           conda-forge/win-64::libhwloc-2.12.2-default_h4379cf1_1000
  libiconv           conda-forge/win-64::libiconv-1.18-hc1393d2_2
  liblapack          conda-forge/win-64::liblapack-3.11.0-6_hf9ab0e9_mkl
  liblzma            conda-forge/win-64::liblzma-5.8.2-hfd05255_0
  libsqlite          conda-forge/win-64::libsqlite-3.52.0-hf5d6505_0
  libwinpthread      conda-forge/win-64::libwinpthread-12.0.0.r4.gg4f2fc60ca-h57928b3_10
  libxml2            conda-forge/win-64::libxml2-2.15.2-h5d26750_0
  libxml2-16         conda-forge/win-64::libxml2-16-2.15.2-h692994f_0
  libzlib            conda-forge/win-64::libzlib-1.3.2-hfd05255_2
  llvm-openmp        conda-forge/win-64::llvm-openmp-22.1.2-h4fa8253_0
  mkl                conda-forge/win-64::mkl-2025.3.1-hac47afa_11
  numpy              conda-forge/win-64::numpy-2.4.3-py312ha3f287d_0
  openssl            conda-forge/win-64::openssl-3.6.1-hf411b9b_1
  packaging          conda-forge/noarch::packaging-26.0-pyhcf101f3_0
  pip                conda-forge/noarch::pip-26.0.1-pyh8b19718_0
  pysocks            conda-forge/noarch::pysocks-1.7.1-pyh09c184e_7
  python             conda-forge/win-64::python-3.12.13-h0159041_0_cpython
  python_abi         conda-forge/noarch::python_abi-3.12-8_cp312
  requests           conda-forge/noarch::requests-2.33.1-pyhcf101f3_0
  setuptools         conda-forge/noarch::setuptools-82.0.1-pyh332efcf_0
  tbb                conda-forge/win-64::tbb-2022.3.0-h3155e25_2
  tk                 conda-forge/win-64::tk-8.6.13-h6ed50ae_3
  tzdata             conda-forge/noarch::tzdata-2025c-hc9c84f9_1
  ucrt               conda-forge/win-64::ucrt-10.0.26100.0-h57928b3_0
  urllib3            conda-forge/noarch::urllib3-2.6.3-pyhd8ed1ab_0
  vc                 conda-forge/win-64::vc-14.3-h41ae7f8_34
  vc14_runtime       conda-forge/win-64::vc14_runtime-14.44.35208-h818238b_34
  vcomp14            conda-forge/win-64::vcomp14-14.44.35208-h818238b_34
  wheel              conda-forge/noarch::wheel-0.46.3-pyhd8ed1ab_0
  win_inet_pton      conda-forge/noarch::win_inet_pton-1.1.0-pyh7428d3b_8
  zstd               conda-forge/win-64::zstd-1.5.7-h534d264_6



Downloading and Extracting Packages:

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Rolling back transaction: done

CondaValueError: Cannot activate environments with '^' in their path from cmd.exe.
Use PowerShell or a caret-free environment path.
()
 

```

**Comment by danyeaw** [comment]
Hey @waqasanjum6, are you able to still reproduce this with the latest canary? If so, could you please open a new issue?
